lib/ClangImporter/DWARFImporter.cpp

```diff
@@ -0,0 +1,178 @@
+/// loading of types is done on demand, this class is effectively empty.
+class DWARFModuleUnit final : public LoadedFile {
+  ~DWARFModuleUnit() = default;
+  ClangImporter::Implementation &owner;
+
+public:
+  DWARFModuleUnit(ModuleDecl &M, ClangImporter::Implementation &owner)
+      : LoadedFile(FileUnitKind::DWARFModule, M), owner(owner) {}
+
+  virtual bool isSystemModule() const override { return false; }
+
+  /// Forwards the request to the ClangImporter, which forwards it to the
+  /// DWARFimporterDelegate.
+  virtual void
+  lookupValue(ModuleDecl::AccessPathTy accessPath, DeclName name,
+              NLKind lookupKind,
+              SmallVectorImpl<ValueDecl *> &results) const override {
+    owner.lookupValueDWARF(accessPath, name, lookupKind, results);
+  }
+
+  virtual TypeDecl *
+  lookupNestedType(Identifier name,
+                   const NominalTypeDecl *baseType) const override {
+    return nullptr;
+  }
+
+  virtual void lookupVisibleDecls(ModuleDecl::AccessPathTy accessPath,
+                                  VisibleDeclConsumer &consumer,
+                                  NLKind lookupKind) const override {}
+
+  virtual void
+  lookupClassMembers(ModuleDecl::AccessPathTy accessPath,
+                     VisibleDeclConsumer &consumer) const override {}
+
+  virtual void
+  lookupClassMember(ModuleDecl::AccessPathTy accessPath, DeclName name,
+                    SmallVectorImpl<ValueDecl *> &decls) const override {}
+
+  void lookupObjCMethods(
+      ObjCSelector selector,
+      SmallVectorImpl<AbstractFunctionDecl *> &results) const override {}
+
+  virtual void
+  getTopLevelDecls(SmallVectorImpl<Decl *> &results) const override {}
+
+  virtual void
+  getDisplayDecls(SmallVectorImpl<Decl *> &results) const override {}
+
+  virtual void
+  getImportedModules(SmallVectorImpl<ModuleDecl::ImportedModule> &imports,
+                     ModuleDecl::ImportFilter filter) const override {}
+
+  virtual void getImportedModulesForLookup(
+      SmallVectorImpl<ModuleDecl::ImportedModule> &imports) const override {};
+
+  virtual void collectLinkLibraries(
+      ModuleDecl::LinkLibraryCallback callback) const override {};
+
+  Identifier
+  getDiscriminatorForPrivateValue(const ValueDecl *D) const override {
+    llvm_unreachable("no private decls in Clang modules");
+  }
+
+  virtual StringRef getFilename() const override { return ""; }
+
+  virtual const clang::Module *getUnderlyingClangModule() const override {
+    return nullptr;
+  }
+
+  static bool classof(const FileUnit *file) {
+    return file->getKind() == FileUnitKind::DWARFModule;
+  }
+  static bool classof(const DeclContext *DC) {
+    return isa<FileUnit>(DC) && classof(cast<FileUnit>(DC));
+  }
+};
+
+ModuleDecl *ClangImporter::Implementation::loadModuleDWARF(
+    SourceLoc importLoc, ArrayRef<std::pair<Identifier, SourceLoc>> path) {
+  // There's no importing from debug info if no importer is installed.
+  if (!DWARFImporter)
+    return nullptr;
```

Comment:
Shouldn't you ask the importer if this module is known? Or are you just assuming any name is potentially valid?